omod-common/src/test/java/org/openmrs/module/webservices/rest/web/api/impl/RestServiceImplTest.java

```diff
@@ -268,4 +271,161 @@ public void getResourceBySupportedClass_shouldReturnTheMostExactMatch() throws E
+		assertFalse(compositeKey2.equals(compositeKey1));
+	}
+	
+	/**
+	 * @verifies return false if given a composite key with different additional key
+	 * @see RestServiceImpl.CompositeSearchHandlerKey#equals(Object)
+	 */
+	@Test
+	public void equals_shouldReturnFalseIfGivenACompositeKeyWithDifferentAdditionalKey() throws Exception {
+		Class compositeSearchHandlerClass = Class
+		        .forName("org.openmrs.module.webservices.rest.web.api.impl.RestServiceImpl$CompositeSearchHandlerKey");
+		Constructor ctor = compositeSearchHandlerClass.getDeclaredConstructors()[0];
+		
+		Object compositeKey1 = ctor.newInstance("v1/order", "default");
+		Object compositeKey2 = ctor.newInstance("v2/order", "custom");
+		assertFalse(compositeKey1.equals(compositeKey2));
+		assertFalse(compositeKey2.equals(compositeKey1));
+	}
+	
+	/**
+	 * @verifies return false if given an object which is not an instanceof this class
+	 * @see RestServiceImpl.CompositeSearchHandlerKey#equals(Object)
+	 */
+	@Test
+	public void equals_shouldReturnFalseIfGivenAnObjectWhichIsNotAnInstanceofThisClass() throws Exception {
+		Class compositeSearchHandlerClass = Class
+		        .forName("org.openmrs.module.webservices.rest.web.api.impl.RestServiceImpl$CompositeSearchHandlerKey");
+		Constructor ctor = compositeSearchHandlerClass.getDeclaredConstructors()[0];
+		
+		Object compositeKey1 = ctor.newInstance("v1/order", "default");
+		assertFalse(compositeKey1.equals("otherClass"));
+	}
+	
+	/**
+	 * @verifies return false if given a composite key with null as supported resource
+	 * @see RestServiceImpl.CompositeSearchHandlerKey#equals(Object)
+	 */
+	@Test
+	public void equals_shouldReturnFalseIfGivenACompositeKeyWithNullAsSupportedResource() throws Exception {
+		Class compositeSearchHandlerClass = Class
+		        .forName("org.openmrs.module.webservices.rest.web.api.impl.RestServiceImpl$CompositeSearchHandlerKey");
+		Constructor ctor = compositeSearchHandlerClass.getDeclaredConstructors()[0];
+		
+		Object compositeKey1 = ctor.newInstance("v1/order", "default");
+		Object compositeKey2 = ctor.newInstance("v2/order", null);
+		assertFalse(compositeKey1.equals(compositeKey2));
+		assertFalse(compositeKey2.equals(compositeKey1));
+	}
+	
+	/**
+	 * @verifies return false if given a composite key with null as additional key
+	 * @see RestServiceImpl.CompositeSearchHandlerKey#equals(Object)
+	 */
+	@Test
+	public void equals_shouldReturnFalseIfGivenACompositeKeyWithNullAsAdditionalKey() throws Exception {
+		Class compositeSearchHandlerClass = Class
+		        .forName("org.openmrs.module.webservices.rest.web.api.impl.RestServiceImpl$CompositeSearchHandlerKey");
+		Constructor ctor = compositeSearchHandlerClass.getDeclaredConstructors()[0];
+		
+		Object compositeKey1 = ctor.newInstance("v1/order", "default");
+		Object compositeKey2 = ctor.newInstance(null, "default");
+		assertFalse(compositeKey1.equals(compositeKey2));
+		assertFalse(compositeKey2.equals(compositeKey1));
```

Comment:
It really looks bad. I saw your discussion with @dkayiwa, but don't you both feel it turned out weird?

You really should be testing the exposed behavior and not if equals works as expected. You will gain more by testing:
```
firstSearchHandler = mock(SearchHandler.class);
firstSearchConfig = new SearchConfig("default", "v1/order", ..., ...);
when(firstSearchHandler.getSearchConfig()).thenReturn(firstSearchConfig);

secondSearchHandler = mock(SearchHandler.class);
secondSearchConfig = new SearchConfig("default", null, ..., ...);
when(secondSearchHandler.getSearchConfig()).thenReturn(secondSearchConfig);

when(restSeviceHelper.getSearchHandlers()).thenReturn(Arrays.asList(firstSearchHandler, secondSearchHandler));

expectedException.expect(InvalidSearchConfigException.class);
expectedException.expectMessage("The resource name for search config should not be null")
restService.getResourceByName("v1/order");
```
When trying to run the test you will actually discover that it fails when calling secondSearchConfig = new SearchConfig("default", null, ..., ...);, which means the violation was caught even earlier.